### PR TITLE
feat: add metrics middleware

### PR DESCRIPTION
**Pull request by frekw**
Since ZIO 2.0 ships with integrated metrics support, I think it makes sense to include middleware for some standard metrics into zio-http.

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (frekw) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: Can be used to override mappings (i.e map certain paths like `/users/1` to `/users/:id`) to avoid having Prometheus label count explode. It would be nice if we could deal with placeholders automatically. Perhaps via the new API work @kitlangton is doing?

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (frekw) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: I think defaulting to Prometheus' default buckets make sense since I suspect most people would be using it, but can be overriden in the constructor.

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (frekw) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: This is needed because gauges can't be incremented or decremented so we have to keep track of the current number of in flight requests ourselves :(

We have to make a tradeoff here: either we use a `Ref` to keep track of the current value and base the Gauge's value off of that. This won't give us readings under concurrency (since the ref and the updating of the gauge aren't atomic).

The alternative is to use `RefM` and update the state and the gauge from within `RefM#update`, but that means that _every request_ needs to acquire two locks for updating the `RefM` which just seems worse to me.

**Comment by tusharmath** [comment]
@frekw This would be a great addition to the ZIO Http ecosystem.
I think it would better to maintain it as separate project, in the same repo.

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: Though I am ok to let people just play around with `zio-http-metrics` initially, 
to gather feedback and focus on performance later. Using a Ref will increase contention between threads for every request. 
What if we use `FiberRef` instead?

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: I think there is quite a lot of logic that's got nothing to do with a middleware. 
How about we create a `Metrics` service that does all this stuff we test that directly instead of doing it via the middleware API.

That way we will later get flexibility to also benchmark the service and see how it performs under stress.

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]:
```suggestion
    def increment(labels: Set[MetricLabel]): URIO[InFlightRequests, Int] = ZIO.serviceWithZIO(_.increment(labels))
    def decrement(labels: Set[MetricLabel]): URIO[InFlightRequests, Int] = ZIO.serviceWithZIO(_.decrement(labels))
```
Shouldn't this be int?

**Comment by frekw** [comment]
> @frekw This would be a great addition to the ZIO Http ecosystem.
I think it would better to maintain it as separate project, in the same repo.

Definitely! What do you think is a good project structure in this case? Should we add an `extras` folder or something similar or should we just add `metrics` to the top level?

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (frekw) [reply zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: I don't really see what that service would do – all helpers essentially extract different metadata about the request or the response to appropriately tag the metrics with. We can definitely move that to its own module if you think it helps readability. 👍

File: zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala
Comment (frekw) [reply zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: Yes, I have [an open ticket](https://github.com/zio/zio/issues/7217) in ZIO to add `Gauge#increment` and `Gauge#decrement` to be able to eliminate the ref entirely. Until it's been decided how that should progress I think we should hold off merging because the performance impact will be too big.

>  What if we use FiberRef instead?

I don't think that'll work unfortunately because the active requests gauge needs to be global for all requests. :/

Comment (frekw) [reply zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala]: This will be fixed by https://github.com/zio/zio/pull/7284 🥳🥳🥳

**Comment by 987Nabil** [comment]
@frekw I would highly recommend to make metric names configurable. Orgs might have naming schemas that a team has to stick to. We have this and I know others as well. And I think making names configurable should not be too complicated.

**Comment by jdegoes** [comment]
@frekw  If you want this to become #1459, I can help shepherd this through to merging. The functionality is most-welcome!

**Comment by frekw** [comment]
@jdegoes now that Gauge.increment/decrement is in core I'd love to take another stab at it 👍

**Comment by frekw** [comment]
@jdegoes I've pushed an updated version that uses the new Gauge.increment/decrement. Temporarily changed the dependency to the snapshot version of ZIO to get access to them but should hold off merging until there's a proper release. It gets a bit repetitive to handle all cases (sucess / error / defect) but nothing too bad.

What I'm struggling a bit with is writing tests that aren't dependent on global state (and execution order) – but I guess that's an issue with testing any logic around metrics, or is there a way to create a scoped, fresh metrics registry solely for testing?

File: zio-http/src/main/scala/zio/http/middleware/Metrics.scala
Comment (frekw) [inline zio-http/src/main/scala/zio/http/middleware/Metrics.scala:29]: Can be used to override mappings (i.e map certain paths like /users/1 to /users/:id) to avoid having Prometheus label count explode (and thus crippling its performance).

**Comment by adamgfraser** [comment]
@frekw I think you can do this with tags.

File: zio-http/src/main/scala/zio/http/middleware/Metrics.scala
Comment (frekw) [inline zio-http/src/main/scala/zio/http/middleware/Metrics.scala]: Not sure about the performance implications of `.ensuring` here – would it be better to just have `observeRequest` accept the `requestLabels` as well and decrement the gauge?

File: project/Dependencies.scala
Comment (frekw) [inline project/Dependencies.scala]: Only for dev, make sure we have 2.0.3 before merge.

File: build.sbt
Comment (frekw) [inline build.sbt]: Only for dev, remove before merge.

File: zio-http/src/main/scala/zio/http/middleware/Metrics.scala
Comment (adamgfraser) [reply zio-http/src/main/scala/zio/http/middleware/Metrics.scala]: You could replace the `tapBoth`, `tapDefect` and `ensuring` with `onExit`.

**Comment by frekw** [comment]
@adamgfraser ah, by accepting something like `extraTags: Set[MetricLabel] = Set.empty`? Clever! 

I'll look at that in the morning, need to jump to bed now! 😴

File: zio-http/src/main/scala/zio/http/middleware/Metrics.scala
Comment (frekw) [reply zio-http/src/main/scala/zio/http/middleware/Metrics.scala]: Ah, TIL! 👍

**Comment by frekw** [comment]
Ok, I think this is good to go @adamgfraser @jdegoes

File: zio-http/src/test/scala/zio/http/middleware/MetricsSpec.scala
Comment (swoogles) [inline zio-http/src/test/scala/zio/http/middleware/MetricsSpec.scala]: Very minor - `Request.method` defaults to `GET`, so that bit can be removed from lines 30-32, 58, and 77.
However, you might prefer the more explicit approach here, since we're tagging things by the method.

**Comment by swoogles** [comment]
Great work here! 👍

**Comment by jdegoes** [comment]
@swoogles Since you also did some work, can you work with @frekw to get this one through?

**Comment by swoogles** [comment]
> @swoogles Since you also did some work, can you work with @frekw to get this one through?

Yep, I've already submitted a PR against his branch with a few small additions. Otherwise I think he nailed it.

**Comment by frekw** [comment]
Sorry @swoogles I didn't have notifications turned on for my fork so I had missed it. Merged now ✅

**Comment by frekw** [comment]
Ok, I've rebased based on the latest changes to `main` and updated the test accordingly. CI should pass now. 👍

**Comment by frekw** [comment]
@jdegoes I noticed master is now using a ZIO snapshot so I've rebased. This should be good to merge now.

**Comment by frekw** [comment]
Whoops, I spoke too early, looks like the Middleware API has changed. I'll adjust.

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
Thank you!

**Pull request merged**
